kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java

```diff
@@ -27,115 +37,169 @@
  * <p>TODO: In future, make chunkInfo read only for more safety.
  *
  * <p>TODO: Is chunk responsible for maintaining it's own metadata?
- *
- * <p>TODO: Move common operations between RO and RW chunk stores into a base class.
  */
 public class ReadOnlyChunkImpl<T> implements Chunk<T> {
 
   private static final Logger LOG = LoggerFactory.getLogger(ReadOnlyChunkImpl.class);
 
-  private final ChunkInfo chunkInfo;
+  private ChunkInfo chunkInfo;
   private LogIndexSearcher<T> logSearcher;
 
-  // TODO: Move this flag into LogStore?.
-  private final boolean readOnly;
-
-  public ReadOnlyChunkImpl(Path dataDirectory, ChunkInfo chunkInfo, MeterRegistry metricsRegistry)
-      throws IOException {
-    this.logSearcher =
-        (LogIndexSearcher<T>)
-            new LogIndexSearcherImpl(LogIndexSearcherImpl.searcherManagerFromPath(dataDirectory));
-    this.chunkInfo = chunkInfo;
-    this.readOnly = true;
-    LOG.info("Created a new read only chunk {}", chunkInfo);
+  private final String chunkId;
+  private final CacheSlotMetadataStore cacheSlotMetadataStore;
+  private final ExecutorService executorService;
+
+  public ReadOnlyChunkImpl(
+      String chunkId,
+      MetadataStoreService metadataStoreService,
+      KaldbConfigs.CacheConfig cacheConfig)
+      throws Exception {
+    this.chunkId = chunkId;
+    this.executorService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat(String.format("readonly-chunk-%s-%%d", chunkId))
+                .build());
+
+    String serverAddress = cacheConfig.getServerConfig().getServerAddress();
+    String slotId = String.format("%s-%s", serverAddress, chunkId);
+
+    String cacheSlotPath = String.format("%s/%s", CACHE_SLOT_STORE_ZK_PATH, slotId);
+    cacheSlotMetadataStore =
+        new CacheSlotMetadataStore(metadataStoreService.getMetadataStore(), cacheSlotPath, true);
+    cacheSlotMetadataStore.addListener(cacheNodeListener());
+
+    CacheSlotMetadata cacheSlotMetadata =
+        new CacheSlotMetadata(
+            METADATA_SLOT_NAME, Metadata.CacheSlotState.FREE, "", Instant.now().toEpochMilli());
+    cacheSlotMetadataStore.create(cacheSlotMetadata);
+
+    LOG.info("Created a new read only chunk {}", chunkId);
   }
 
-  @Override
-  public void addMessage(T message) {
-    throw new ReadOnlyChunkInsertionException(
-        String.format("Chunk %s is a read only chunk", chunkInfo));
+  private KaldbMetadataStoreChangeListener cacheNodeListener() {
+    return () -> {
+      CacheSlotMetadata cacheSlotMetadata = cacheSlotMetadataStore.getCached().get(0);
```

Comment:
looks like everywhere we do a `get(0)` ? I guess that's because we have to conform to how the `CachedMetadataStore` hierarchy is?

So like every cache node gets it's znode and whenever the first entry  under it updates to evict/assigned etc we do something

looks fine i guess

Reply:
will there be N top level znodes for every chunk within a cache node?

Reply:
The current design is such that the `CachedMetadataStoreImpl` caches all nested nodes. As we know there exist only a single nested node (`SLOT` in this case) we can safely just get the first item in the list. 

We have a todo to refactor the current metadata store constructors, such that we can initialize the store with data on the top znode, and not have to put the data on a nested node. That will simplify access to these znodes once that's refactored.